cmd/snap-update-ns/utils.go

```diff
@@ -667,3 +667,92 @@ func createWritableMimic(dir, neededBy string, sec *Secure) ([]*Change, error) {
 	}
 	return changes, nil
 }
+
+// PathIterator traverses through segments (directories and files) of some
+// path. The filesystem is never consulted, traversal is done purely in memory.
+//
+// The iterator is useful in implementing secure traversal of absolute paths
+// using the common idiom of opening the root directory followed by a chain of
+// openat calls.
+//
+// A simple example on how to use the iterator:
+// ```
+// pi := NewPathIterator(path)
+// for pi.Next() {
+//    // pi.PathSoFar() or pi.Segment()
+// }
+// ```
+type PathIterator struct {
+	path        string
+	left, right int
+}
+
+// NewPathIterator returns an iterator for traversing the given path.
+// The path is passed through filepath.Clean automatically.
+func NewPathIterator(path string) *PathIterator {
+	return &PathIterator{path: filepath.Clean(path)}
+}
+
+// Path returns the path being traversed.
+func (pi *PathIterator) Path() string {
```

Comment:
You could drop this method and just use:
```
type PathIterator struct{
    Path string
    ....
}
```
Then the callers could just create the struct on their stack rather than be forced to call `NewPathIterator()`.

Reply:
but then anyone could corrupt the state by meddling with path.